### PR TITLE
feat: telemetry

### PR DESCRIPTION
**Pull request by odysa**
I hereby agree to the terms of the [Singularity Data, Inc. Contributor License Agreement](https://gist.github.com/skyzh/0663682a70b0edde7ae991492f2314cb#file-s9y_cla).

## What's changed and what's your intention?

Telemetry enables RisingWave to analyze performance from running instances.

## Checklist

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [X] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features).
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Documentation

To disable telemetry, either set in `./src/config/<your-config>.yaml`
```toml
[server]
telemetry_enabled = false
```
or set an environment variable
```bash
export ENABLE_TELEMETRY=false
```
or set system params via psql
```
ALTER SYSTEM SET telemetry_enabled = false
```
## Refer to a related PR or issue link (optional)
#7105

### CODE REVIEW

File: .cargo/config.toml [hunk 1]
```diff
@@ -42,4 +42,6 @@ rustflags = [
     "-Wfuture_incompatible",
     "-Wnonstandard_style",
     "-Wrust_2018_idioms",
+    "-C",
```
Comment (jon-chuang) [inline .cargo/config.toml:45]: why is this necessary?

Comment (odysa) [reply .cargo/config.toml:45]: see #7375

File: src/common/src/telemetry/report.rs [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+        &self,
+        tracking_id: String,
+        session_id: String,
+        up_time: u64,
+    ) -> Result<impl TelemetryReport>;
+
+    fn report_type(&self) -> &str;
+}
+
+pub fn start_telemetry_reporting<F, I>(
```
Comment (fuyufjh) [inline src/common/src/telemetry/report.rs:43]: So every node is looping to proactivaly fetch telemetry from meta service. This looks like an uncessary RPC burden for the meta service. Is it possible to issue the "start_telemetry_reporting" request from Meta service?

File: src/compute/src/telemetry.rs [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+        ))
+    }
+
+    fn report_type(&self) -> &str {
+        "compute"
+    }
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+pub(crate) struct ComputeTelemetryReport {
```
Comment (fuyufjh) [inline src/compute/src/telemetry.rs:51]: Will we add more metrics later? I think you may pick some key metrics from our Grafana dashboard. May also ask @lmatz

File: src/common/src/telemetry/mod.rs [hunk 4]
```diff
@@ -0,0 +1,225 @@
...
+#[derive(Debug, Serialize, Deserialize, PartialEq, Eq)]
+pub enum TelemetryNodeType {
+    Meta,
+    Compute,
+    Frontend,
+    Compactor,
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+pub struct TelemetryReportBase {
```
Comment (fuyufjh) [inline src/common/src/telemetry/mod.rs:45]: Consider system metrics as well. e.g. CPU, memory, total memory in machines, # cpu cores in machine, ...

File: src/storage/compactor/Cargo.toml [hunk 5]
```diff
@@ -15,6 +15,7 @@ ignored = ["workspace-hack"]
 normal = ["workspace-hack"]
 
 [dependencies]
+anyhow = "1.0.68"
```
Comment (fuyufjh) [inline src/storage/compactor/Cargo.toml:18]: Recommend to match major version only

```suggestion
anyhow = "1"
```

File: src/storage/compactor/Cargo.toml [hunk 6]
```diff
@@ -26,6 +27,9 @@ risingwave_object_store = { path = "../../object_store" }
 risingwave_pb = { path = "../../prost" }
 risingwave_rpc_client = { path = "../../rpc_client" }
 risingwave_storage = { path = "../../storage" }
+risingwave_tracing = { path = "../../tracing" }
+serde = { version = "1", features = ["derive"] }
+serde_json = "1.0.91"
```
Comment (fuyufjh) [inline src/storage/compactor/Cargo.toml:32]: ditto.

```suggestion
serde_json = "1"
```

**Review by fuyufjh — COMMENTED** [review]
Rest LGTM

File: src/common/src/config.rs [hunk 7]
```diff
@@ -187,6 +187,9 @@ pub struct ServerConfig {
     /// 0 = close metrics
     /// >0 = open metrics
     pub metrics_level: u32,
+
+    #[serde(default = "default::server::telemetry_enabled")]
+    pub telemetry_enabled: bool,
```
Comment (hzxa21) [inline src/common/src/config.rs:192]: I think we can make `telemetry_enabled` as part of the cluster-level system parameter https://github.com/risingwavelabs/risingwave/blob/c4516dea1366d9df06ccf7e28ef2bdabf6fb1019/src/common/src/system_param.rs#L28

This makes it possible for user to enable/disable telemetry in runtime and nodes in the same cluster can have the same `telemetry_enabled` config.

Comment (hzxa21) [reply src/common/src/config.rs:192]: Actually, if we also make `tracking_id` as part of the system parameter, we may not need GetTelemetryInfo anymore.

File: src/common/src/telemetry/report.rs [hunk 8]
```diff
@@ -0,0 +1,124 @@
...
+                    }
+                };
+
+            if should_kill_telemetry {
+                tracing::info!("Telemetry is shutdown due to meta resp");
+                return;
+            }
+
+            // wait for the next interval, do not exit current thread
+            if !telemetry_enabled {
```
Comment (hzxa21) [inline src/common/src/telemetry/report.rs:89]: `TELEMETRY_REPORT_INTERVAL` is 2hr. That means if user turns off telemetry, it can take as long as 2hr to take effect, which is too long. Ideally, telemetry should be able to be disabled within a few seconds/mins. Again, if we use system parameter to control this, the change can be made in a more timely manner because we can use async notification when telemetry is enabled/disabled.

File: src/common/src/telemetry/mod.rs [hunk 9]
```diff
@@ -0,0 +1,225 @@
...
+use anyhow::{anyhow, Result};
+use hyper::StatusCode;
+use serde::{Deserialize, Serialize};
+use sysinfo::{System, SystemExt};
+
+use crate::util::resource_util::cpu::total_cpu_available;
+use crate::util::resource_util::memory::{total_memory_available_bytes, total_memory_used_bytes};
+
+/// Url of telemetry backend
+pub const TELEMETRY_REPORT_URL: &str = "http://localhost:8080/v1/report";
```
Comment (hzxa21) [inline src/common/src/telemetry/mod.rs:28]: Should this be configurable?

File: src/meta/src/telemetry.rs [hunk 10]
```diff
@@ -0,0 +1,165 @@
...
+}
+
+/// This function spawns a new tokio task to report telemetry.
+/// It creates a channel for killing itself and a join handle to the spawned task.
+/// It then creates an interval of `TELEMETRY_REPORT_INTERVAL` seconds and checks if telemetry is
+/// enabled. If it is, it gets or creates a tracking ID from the meta store,
+/// creates a `TelemetryReport` object with system data, uptime, timestamp, and tracking ID.
+/// Finally, it posts the report to `TELEMETRY_REPORT_URL`.
+/// If an error occurs at any point in the process, it logs an error message.
+pub async fn start_meta_telemetry_reporting(
```
Comment (hzxa21) [inline src/meta/src/telemetry.rs:48]: `start_meta_telemetry_reporting` seems to be very similar to `start_telemetry_reporting`. Wonder whether we can resue some common logics to avoid code duplication.

File: src/common/src/config.rs (hunk 7 above)
Comment (odysa) [reply src/common/src/config.rs:192]: IMO, if we move `tracking_id` to system params, it will be configurable while it is not a config field.
 It's a good idea to move `telemetry_enabled` to system params.

**Comment by odysa** [comment]
Integration with system parameters has been done. Now we can shutdown telemetry by `ALTER SYSTEM SET telemetry_enabled = false`.
`tracking_id` is only fetched when a node starts.
Refactor some code.

File: src/meta/src/telemetry.rs [hunk 11]
```diff
@@ -0,0 +1,159 @@
...
+            .await
+            .map(|id| id.to_string())?;
+
+        Ok(tracking_id)
+    }
+}
+
+/// fetch or create a `tracking_id` from etcd
+async fn get_or_create_tracking_id(meta_store: Arc<impl MetaStore>) -> Result<Uuid, anyhow::Error> {
+    match meta_store.get_cf(TELEMETRY_CF, TELEMETRY_KEY).await {
```
Comment (hzxa21) [inline src/meta/src/telemetry.rs:82]: Should we include tracking id in our backup process? (See https://github.com/odysa/risingwave/blob/4a5b39e978922f3318579b336239993d46eea659/src/meta/src/backup_restore/meta_snapshot_builder.rs#L47)

File: src/common/src/telemetry/mod.rs [hunk 12]
```diff
@@ -0,0 +1,193 @@
...
+use sysinfo::{System, SystemExt};
+
+use crate::util::resource_util::cpu::total_cpu_available;
+use crate::util::resource_util::memory::{total_memory_available_bytes, total_memory_used_bytes};
+
+/// Url of telemetry backend
+pub const TELEMETRY_REPORT_URL: &str = "https://telemetry.risingwave.dev/api/v1/report";
+
+/// Telemetry reporting interval in seconds, 24h
+pub const TELEMETRY_REPORT_INTERVAL: u64 = 24 * 60 * 60;
```
Comment (hzxa21) [inline src/common/src/telemetry/mod.rs:31]: 1day seems a bit too long for me. How about 3hr or 6hr?

File: src/compute/src/server.rs [hunk 13]
```diff
@@ -316,6 +322,21 @@ pub async fn compute_node_serve(
...
 
+    let telemetry_manager = TelemetryManager::new(
+        system_params_manager.watch_params(),
+        Arc::new(meta_client.clone()),
+        Arc::new(ComputeTelemetryCreator::new()),
+    );
+
+    // if the toml config file or env variable disables telemetry, do not watch system params change
+    // because if any of configs disable telemetry, we should never start it
+    if config.server.telemetry_enabled && telemetry_env_enabled() && telemetry_enabled {
```
Comment (hzxa21) [inline src/compute/src/server.rs:333]: I think we should stll `watch_params_change` if the system param `telemetry_enabled` is false given that it can be changed in runtime.

File: src/frontend/src/session.rs [hunk 14]
```diff
@@ -261,6 +268,26 @@ impl FrontendEnv {
...
+
+        let telemetry_manager = TelemetryManager::new(
+            system_params_manager.watch_params(),
+            Arc::new(meta_client.clone()),
+            Arc::new(FrontendTelemetryCreator::new()),
+        );
+
+        // if the toml config file or env variable disables telemetry, do not watch system params
+        // change because if any of configs disable telemetry, we should never start it
+        if config.server.telemetry_enabled && telemetry_env_enabled() && telemetry_enabled {
```
Comment (hzxa21) [inline src/frontend/src/session.rs:280]: ditto

File: src/meta/src/rpc/server.rs [hunk 15]
```diff
@@ -538,6 +545,22 @@ pub async fn start_service_as_election_leader<S: MetaStore>(
...
+    let local_system_params_manager = LocalSystemParamsManager::new(system_params_reader.clone());
+
+    let mgr = TelemetryManager::new(
+        local_system_params_manager.watch_params(),
+        Arc::new(MetaTelemetryInfoFetcher::new(meta_store.clone())),
+        Arc::new(MetaReportCreator::new()),
+    );
+
+    // May start telemetry reporting
+    if let MetaBackend::Etcd = meta_store.meta_store_type() && system_params_reader.telemetry_enabled() && env.opts.telemetry_enabled{
```
Comment (hzxa21) [inline src/meta/src/rpc/server.rs:557]: ditto

File: src/storage/compactor/src/server.rs [hunk 16]
```diff
@@ -162,6 +169,20 @@ pub async fn compactor_serve(
...
     ];
 
+    let telemetry_manager = TelemetryManager::new(
+        system_params_manager.watch_params(),
+        Arc::new(meta_client.clone()),
+        Arc::new(CompactorTelemetryCreator::new()),
+    );
+    // if the toml config file or env variable disables telemetry, do not watch system params change
+    // because if any of configs disable telemetry, we should never start it
+    if config.server.telemetry_enabled && telemetry_env_enabled() && telemetry_enabled {
```
Comment (hzxa21) [inline src/storage/compactor/src/server.rs:179]: ditto

File: src/storage/compactor/Cargo.toml [hunk 17]
```diff
@@ -26,6 +27,9 @@ risingwave_object_store = { path = "../../object_store" }
 risingwave_pb = { path = "../../prost" }
 risingwave_rpc_client = { path = "../../rpc_client" }
 risingwave_storage = { path = "../../storage" }
+risingwave_tracing = { path = "../../tracing" }
```
Comment (hzxa21) [inline src/storage/compactor/Cargo.toml:30]: where do we need this?

**Review by hzxa21 — APPROVED** [review]
Rest LGTM.

Great work! Thanks for the PR.

File: src/compute/src/server.rs (hunk 13 above)
Comment (odysa) [reply src/compute/src/server.rs:333]: So, we always start watching system_params If both the config and the environment variable are true.
```rust
if config.server.telemetry_enabled && telemetry_env_enabled(){
   // start watching 
}
```

**Comment by odysa** [comment]
> how can we test the telemetry logics in the future 

A mock HTTP server in e2e tests may work. I am looking into it to see how to start a mock server in e2e tests awaiting telemetry reports.

File: src/common/src/telemetry/mod.rs (hunk 12 above)
Comment (odysa) [reply src/common/src/telemetry/mod.rs:31]: 6hr LGTM

**Comment by hzxa21** [comment]
Just realize that we also need to disable telemetry reporting in the following places:
- RiseDev. Example on how to set a global env in risedev: https://github.com/risingwavelabs/risingwave/pull/7828
- Demo integration tests. We can pass [env file](https://docs.docker.com/compose/environment-variables/set-environment-variables/#substitute-with---env-file) for docker compose in the [python script](https://github.com/risingwavelabs/risingwave/blob/23bc699d1c1ffe54f4bc42bb4cfe13baf95545f8/integration_tests/scripts/run_demos.py#L28)

**Review by wenym1 — APPROVED** [review]

**Review by wenym1 — APPROVED** [review]
